src/MOIWrapper.jl

```diff
@@ -0,0 +1,373 @@
+function _allocateconstraint!(cone::Cone, f, s::MOI.SecondOrderCone)
+    push!(cone.qa, s.dimension)
+    ci = cone.q
+    cone.q += _dim(s)
+    ci
+end
+constroffset(cone::Cone, ci::CI{<:MOI.AbstractFunction, <:MOI.PositiveSemidefiniteConeTriangle}) = cone.f + cone.l + cone.q + ci.value
+function _allocateconstraint!(cone::Cone, f, s::MOI.PositiveSemidefiniteConeTriangle)
+    push!(cone.sa, s.dimension)
+    ci = cone.s
+    cone.s += _dim(s)
+    ci
+end
+constroffset(cone::Cone, ci::CI{<:MOI.AbstractFunction, <:MOI.ExponentialCone}) = cone.f + cone.l + cone.q + cone.s + ci.value
+function _allocateconstraint!(cone::Cone, f, s::MOI.ExponentialCone)
+    ci = 3cone.ep
+    cone.ep += 1
+    ci
+end
+constroffset(optimizer::SCSOptimizer, ci::CI) = constroffset(optimizer.cone, ci::CI)
+MOIU.canallocateconstraint(::SCSOptimizer, ::Type{<:SF}, ::Type{<:SS}) = true
+function MOIU.allocateconstraint!(optimizer::SCSOptimizer, f::F, s::S) where {F <: MOI.AbstractFunction, S <: MOI.AbstractSet}
+    CI{F, S}(_allocateconstraint!(optimizer.cone, f, s))
+end
+
+# Vectorized length for matrix dimension n
+sympackedlen(n) = (n*(n+1)) >> 1
+# Matrix dimension for vectorized length n
+sympackeddim(n) = div(isqrt(1+8n) - 1, 2)
+trimap(i::Integer, j::Integer) = i < j ? trimap(j, i) : div((i-1)*i, 2) + j
+trimapL(i::Integer, j::Integer, n::Integer) = i < j ? trimapL(j, i, n) : i + div((2n-j) * (j-1), 2)
+function _sympackedto(x, n, mapfrom, mapto)
+    @assert length(x) == sympackedlen(n)
+    y = similar(x)
+    for i in 1:n, j in 1:i
+        y[mapto(i, j)] = x[mapfrom(i, j)]
+    end
+    y
+end
+sympackedLtoU(x, n=sympackeddim(length(x))) = _sympackedto(x, n, (i, j) -> trimapL(i, j, n), trimap)
+sympackedUtoL(x, n=sympackeddim(length(x))) = _sympackedto(x, n, trimap, (i, j) -> trimapL(i, j, n))
+
+function sympackedUtoLidx(x::AbstractVector{<:Integer}, n)
+    y = similar(x)
+    map = sympackedLtoU(1:sympackedlen(n), n)
+    for i in eachindex(y)
+        y[i] = map[x[i]]
+    end
+    y
+end
+
+
+# Build constraint matrix
+scalecoef(rows, coef, minus, s) = minus ? -coef : coef
+scalecoef(rows, coef, minus, s::Union{MOI.LessThan, Type{<:MOI.LessThan}, MOI.Nonpositives, Type{MOI.Nonpositives}}) = minus ? coef : -coef
+function _scalecoef(rows, coef, minus, d, rev)
+    scaling = minus ? -1 : 1
+    scaling2 = rev ? scaling / sqrt(2) : scaling * sqrt(2)
+    output = copy(coef)
+    diagidx = IntSet()
+    for i in 1:d
+        push!(diagidx, trimap(i, i))
+    end
+    for i in 1:length(output)
+        if rows[i] in diagidx
+            output[i] *= scaling
+        else
+            output[i] *= scaling2
+        end
+    end
+    output
+end
+function scalecoef(rows, coef, minus, s::MOI.PositiveSemidefiniteConeTriangle)
+    _scalecoef(rows, coef, minus, s.dimension, false)
+end
+function scalecoef(rows, coef, minus, ::Type{MOI.PositiveSemidefiniteConeTriangle})
```

Comment:
Why are there separate methods for `Type{MOI.PositiveSemidefiniteConeTriangle}` and `MOI.PositiveSemidefiniteConeTriangle`?